File: internal/handlers/media.go
package handlers

import (
	"net/http"
	"net/url"
	"strconv"

	"ramiro-uziel/vault/internal/apperr"
	"ramiro-uziel/vault/internal/auth"
	"ramiro-uziel/vault/internal/httputil"
	"ramiro-uziel/vault/internal/middleware"
)

type MediaHandler struct {
	config auth.Config
}

func NewMediaHandler(config auth.Config) *MediaHandler {
	return &MediaHandler{config: config}
}

func (h *MediaHandler) StreamURL(w http.ResponseWriter, r *http.Request) error {
	userID, err := httputil.RequireUserID(r)
	if err != nil {
		return apperr.NewUnauthorized("unauthorized")
	}

	trackID := r.PathValue("id")
	if trackID == "" {
		return apperr.NewBadRequest("track id is required")
	}

	query := url.Values{}
	query.Set("user_id", strconv.Itoa(userID))

	quality := r.URL.Query().Get("quality")
	if quality != "" {
		query.Set("quality", quality)
	}
	versionID := r.URL.Query().Get("version_id")
	if versionID != "" {
		query.Set("version_id", versionID)
	}

	path := "/api/stream/" + trackID
	url, err := middleware.BuildSignedURL("", path, query, h.config.SignedURLSecret, h.config.SignedURLExpiration)
	if err != nil {
		return apperr.NewInternal("failed to build signed url", err)
	}

	return httputil.OKResult(w, map[string]string{"url": url})
}

func (h *MediaHandler) ProjectCoverURL(w http.ResponseWriter, r *http.Request) error {
	userID, err := httputil.RequireUserID(r)
	if err != nil {
		return apperr.NewUnauthorized("unauthorized")
	}

	projectID := r.PathValue("id")
	if projectID == "" {
		return apperr.NewBadRequest("project id is required")
	}

	query := url.Values{}
	query.Set("user_id", strconv.Itoa(userID))

	size := r.URL.Query().Get("size")
	if size != "" {
		query.Set("size", size)
	}

	path := "/api/projects/" + projectID + "/cover"
	url, err := middleware.BuildSignedURL("", path, query, h.config.SignedURLSecret, h.config.SignedURLExpiration)
	if err != nil {
		return apperr.NewInternal("failed to build signed url", err)
	}

	return httputil.OKResult(w, map[string]string{"url": url})
}

File: frontend/src/api/media.ts
import { get } from './client'

export async function getStreamUrl(
	trackId: string,
	params?: { quality?: string; versionId?: number | null }
): Promise<{ url: string }> {
	const query = new URLSearchParams()
	if (params?.quality) {
		query.set('quality', params.quality)
	}
	if (params?.versionId) {
		query.set('version_id', String(params.versionId))
	}
	const suffix = query.toString() ? `?${query.toString()}` : ''
	return get<{ url: string }>(`/api/media/stream/${trackId}${suffix}`)
}

export async function getProjectCoverUrl(
	projectId: string,
	params?: { size?: string }
): Promise<{ url: string }> {
	const query = new URLSearchParams()
	if (params?.size) {
		query.set('size', params.size)
	}
	const suffix = query.toString() ? `?${query.toString()}` : ''
	return get<{ url: string }>(`/api/media/projects/${projectId}/cover${suffix}`)
}
